### PR TITLE
extend the definition of q-binomial coefficients to (k > 0, n < 0)

### PR DESCRIPTION
**Pull request by fchapoton**
This is extending the domain of the q-binomial coefficients.

This is done by polynomiality wrt n, for k >= 0.

### :memo: Checklist

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: src/sage/combinat/q_analogues.py [hunk 1]
```diff
@@ -328,6 +330,11 @@ def q_binomial(n, k, q=None, algorithm='auto'):
         from sage.rings.polynomial.polynomial_element import Polynomial
         is_polynomial = isinstance(q, Polynomial)
 
+    if n >= 0:
+        k = min(n - k, k)  # Pick the smallest k
+    else:
+        return (-1)**k * q**(k * n - (k * k - k) // 2) * q_binomial(-n + k - 1, k, q=q)
```
Comment (tscrim) [inline src/sage/combinat/q_analogues.py]:
```suggestion
    if n < 0:
        return (-1)**k * q**(k * n - (k * k - k) // 2) * q_binomial(-n + k - 1, k, q=q)

    k = min(n - k, k)  # Pick the smallest k        
```
More of a stylistic thing, but I find this easier to follow the logic.

**Review by tscrim — CHANGES_REQUESTED** [review]
LGTM overall. I have one minor code stylistic change, and one question. Should the default ring for the $n < 0$ case be $\mathbb{Z}[q,q^{-1}]$ instead of the fraction field of $\mathbb{Z}[q]$? This has properties closer to the polynomial ring and arithmetic is faster. What do you think?

**Comment by fchapoton** [comment]
yes, right, this is natural. Done, also for `q_int`.

**Review by tscrim — APPROVED** [review]
Thanks. LGTM.

(Personally, I am not a fan of `polygen` as it is less explicit, but that's just my preference.)

**Pull request merged**